src/commands/templates.rs

```diff
@@ -181,18 +186,26 @@ impl Install {
 
 impl Upgrade {
     pub async fn run(&self) -> Result<()> {
+        let template_manager = TemplateManager::try_default()?;
+        let installed: Vec<String> = template_manager
+            .list()
+            .await?
+            .templates
+            .into_iter()
+            .map(|s| s.id().to_owned())
+            .collect();
         if self.git.is_some() {
             // This is equivalent to `install --update`
             let install = Install {
                 git: self.git.clone(),
                 branch: self.branch.clone(),
                 dir: None,
                 update: true,
+                id: installed,
```

Comment:
Doesn't this mean that if we add a new template to the Spin repo, existing users will never get it?  That seems regrettable.

Reply:
Yes but the issue that I see with the current approach is that, If I install all the templates from Spin repo and then decide I do not want to write zig or c code and delete the templates and then upgrade, I end up getting all the templates installed again.

Reply:
Yes but deleting C doesn't mean I want to miss out on Befunge when I've been waiting so long for it.  Worse still, with this, even if I _never deleted anything_ I will _still_ never get Befunge, because this implementation _locks me in_ to my current set of templates.

I'm fully in favour of respecting deletes or "I only ever asked for http-rust" across an upgrade - but we must have an answer for that _first._  We need to write down the desired (and undesired) behaviours, and then _design_ something that will behave correctly, and I fear that will have to be rather more complicated and nuanced than this simple initial approach (which I realise is a draft for feedback, but this is the feedback _grin_).

Reply:
I agree with this needing discussion. My take on this is that `upgrade` only upgrades what I have. If I need something new that I have never had I would have to `install` first. I can definitely be convinced otherwise. 

I am happy to remove the changes to the `upgrade`  part and leave the behavior as it is today if that is a preferred approach.

Reply:
Yes, the trouble is that historically what I have had is a _repo_ rather than a set of individual templates.  (For better or for worse!) 

As for what to do in this PR... the thing is, I agree that if someone has said "I only want http-rust" then an upgrade should only give them http-rust.  The current behaviour is definitely not the preferred approach!  So I am not sure that we can punt on this unfortunately...

Reply:
I have given it a little more thought and to me, the more I am convinced `upgrade` should not bring in anything new. Even though it has historically done so, I think using the command `spin templates install --update ...` is what I seem to recall in most places so this should not have too much impact potentially (?).